### PR TITLE
rgw: cls_fifo-based BILog, the updater part

### PR DESCRIPTION
**Pull request by rzarzynski**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_rados.h [hunk 1]
```diff
@@ -1361,8 +1361,8 @@ class RGWRados
 
   int cls_obj_prepare_op(BucketShard& bs, RGWModifyOp op, string& tag, rgw_obj& obj, uint16_t bilog_flags, optional_yield y, rgw_zone_set *zones_trace = nullptr);
   int cls_obj_complete_op(BucketShard& bs, const rgw_obj& obj, RGWModifyOp op, string& tag, int64_t pool, uint64_t epoch,
-                          rgw_bucket_dir_entry& ent, RGWObjCategory category, list<rgw_obj_index_key> *remove_objs, uint16_t bilog_flags, rgw_zone_set *zones_trace = nullptr);
-  int cls_obj_complete_add(BucketShard& bs, const rgw_obj& obj, string& tag, int64_t pool, uint64_t epoch, rgw_bucket_dir_entry& ent,
+                          const rgw_bucket_dir_entry& ent, RGWObjCategory category, list<rgw_obj_index_key> *remove_objs, uint16_t bilog_flags, rgw_zone_set *zones_trace = nullptr);
```
Comment (adamemerson) [inline src/rgw/rgw_rados.h:1306]: A lot of stuff being mutable that didn't need to be was annoying me, so I appreciate this.

File: src/rgw/rgw_service.cc [hunk 2]
```diff
@@ -53,6 +53,12 @@ int RGWServices_Def::init(CephContext *cct,
   bucket_sobj = std::make_unique<RGWSI_Bucket_SObj>(cct);
   bucket_sync_sobj = std::make_unique<RGWSI_Bucket_Sync_SObj>(cct);
   bi_rados = std::make_unique<RGWSI_BucketIndex_RADOS>(cct);
+  // Q: will we have the bucket layout available here?
+  // A: metadata log uses the `neorados` along the path to determine
+  //    which backend (fifo / log) should be used. Would it be enough
+  //    BILog as well?
+  //    Liekly we would need to introduce use `bilog_rados->init()`
+  //    as `datalog_rados` does with its `start()`. 
   bilog_rados = std::make_unique<RGWSI_BILog_RADOS>(cct);
```
Comment (adamemerson) [inline src/rgw/rgw_service.cc:62]: I appreciate this elucidation of design inline.

File: src/rgw/rgw_service.cc [hunk 3]
```diff
@@ -53,6 +53,12 @@ int RGWServices_Def::init(CephContext *cct,
   bucket_sobj = std::make_unique<RGWSI_Bucket_SObj>(cct);
   bucket_sync_sobj = std::make_unique<RGWSI_Bucket_Sync_SObj>(cct);
   bi_rados = std::make_unique<RGWSI_BucketIndex_RADOS>(cct);
+  // Q: will we have the bucket layout available here?
+  // A: metadata log uses the `neorados` along the path to determine
```
Comment (adamemerson) [inline src/rgw/rgw_service.cc:57]: We won't be using neorados in this run, though. The current plan is to use the interface in cls_fifo_legacy for ease of backporting, and then it can be updated to neorados for better Aio support later.

File: src/rgw/rgw_rados.cc [hunk 4]
```diff
@@ -6845,13 +6845,30 @@ int RGWRados::block_while_resharding(RGWRados::BucketShard *bs,
   return -ERR_BUSY_RESHARDING;
 }
 
+template <class CLSRGWBucketModifyOpT, class F, class... Args>
+int RGWRados::with_bilog(F&& on_flushed, Args&&... args)
+{
+  constexpr bool is_inindex = true;
+  if (is_inindex) {
+    return std::move(on_flushed)(
```
Comment (adamemerson) [inline src/rgw/rgw_rados.cc:6869]: Minor nit: Use std::forward here, not std::move.

File: src/rgw/rgw_rados.cc [hunk 5]
```diff
@@ -828,16 +828,8 @@ struct complete_op_data {
   AioCompletion *rados_completion{nullptr};
   int manager_shard_id{-1};
   RGWIndexCompletionManager *manager{nullptr};
+  std::function<void(librados::ObjectWriteOperation& o)> bi_update;
```
Comment (adamemerson) [inline src/rgw/rgw_rados.cc:832]: If you don't need to copy it, you might prefer `fu2::unique_function`, since then our functions can include unique_ptrs and other uncopyable stuff.

(`unique_function` will not be making it into C++20 because they couldn't agree on whether to rename it to `any_invocable`, apparently. Sheesh.)

File: src/rgw/rgw_rados.cc [hunk 6]
```diff
@@ -1011,42 +998,45 @@ static void obj_complete_cb(completion_t cb, void *arg)
 }
 
 
-void RGWIndexCompletionManager::create_completion(const rgw_obj& obj,
-                                                  RGWModifyOp op, string& tag,
+static std::list<cls_rgw_obj_key>
+maybe_copy_remove_objs(std::list<cls_rgw_obj_key>* const remove_objs)
```
Comment (adamemerson) [inline src/rgw/rgw_rados.cc:998]: If possible without gigantic scope creep, I would prefer to get rid of std::lists.

(But obviously getting things working quickly is a bigger priority.)

File: src/rgw/rgw_rados.cc [hunk 7]
```diff
@@ -8294,24 +8272,32 @@ int RGWRados::cls_obj_complete_op(BucketShard& bs, const rgw_obj& obj, string& t
...
       bi_updater.complete_op(o, ver, dir_meta, remove_objs);
-      complete_op_data *arg;
-      index_completion_manager->create_completion(std::move(bi_updater),
-                                                  obj,
-                                                  ver,
-                                                  dir_meta,
-                                                  remove_objs,
-                                                  &arg);
+
+      // TODO: it would be really cool to knock the completion manager
```
Comment (adamemerson) [inline src/rgw/rgw_rados.cc:8344]: This is the purpose of the async-work-to-be.

**Review by adamemerson — COMMENTED** [review]
I approve of this so far!

**Comment by rzarzynski** [comment]
Continuing in the [`wip-rgw-bilog-in-cls_fifo_reader`](https://github.com/ceph/ceph/compare/master...rzarzynski:wip-rgw-bilog-in-cls_fifo_reader) branch.
